File: src/db/engines/mssql.py
"""MSSQL (SQL Server) engine implementation using aioodbc."""

from typing import Any, Dict, List, Optional

try:
    HAS_AIOODBC = True
except ImportError:
    HAS_AIOODBC = False
    create_async_engine: Any = None
    AsyncEngine: Any = None
    text: Any = None

# ─────────────────────────────────────────────────────────────────────────────
from sqlalchemy import text
from sqlalchemy.ext.asyncio import create_async_engine

from config.schema import DatabaseDefConfig
from db.engines.base import (
    ColumnInfo,
    DatabaseEngine,
    ForeignKeyInfo,
    QueryResult,
    TableInfo,
)
from encoding.proto_utils import _encode_value
from generated.axiom.v1 import db_pb2

# Helper Functions
# ─────────────────────────────────────────────────────────────────────────────


def _normalize_uri(raw_uri: str) -> str:
    """Safely transposes human-readable schemes to driver-specific representations."""
    if raw_uri.startswith("mssql://"):
        return raw_uri.replace("mssql://", "mssql+aioodbc://")
    if raw_uri.startswith("sqlserver://"):
        return raw_uri.replace("sqlserver://", "mssql+aioodbc://")
    return raw_uri


def _is_mutation_query(sql: str) -> bool:
    """Determines if the raw payload enforces write locks or mutates schema/data."""
    return (
        sql.strip()
        .upper()
        .startswith(
            ("INSERT", "UPDATE", "DELETE", "TRUNCATE", "DROP", "CREATE", "ALTER")
        )
    )


async def _execute_mutation(conn, statement, params: dict) -> QueryResult:
    """Dispatches a mutation and forces an explicit commit."""
    result = await conn.execute(statement, params)
    await conn.commit()
    return QueryResult(affected_rows=result.rowcount)


async def _execute_read(
    conn, statement, params: dict, return_format: str
) -> QueryResult:
    """Resolves standard non-mutating statements into explicit mapping lists."""
    result = await conn.execute(statement, params)
    columns = list(result.keys()) if result.keys() else []

    if not result.returns_rows:
        # Commit dynamically to free isolation level row locks gracefully
        await conn.commit()
        return QueryResult(columns=columns, rows=[], affected_rows=result.rowcount)

    if return_format == "protobuf":
        pb = db_pb2.ExecuteQueryResponse()
        pb.columns.extend(columns)

        for row in result:
            pb_row = pb.rows.add()
            for val in row:
                _encode_value(pb_row.values.add(), val)

        return QueryResult(columns=columns, affected_rows=result.rowcount, proto_msg=pb)
    else:
        rows = [dict(row._mapping) for row in result]
        return QueryResult(columns=columns, rows=rows, affected_rows=result.rowcount)


# ─────────────────────────────────────────────────────────────────────────────
# MSSQL Driver
# ─────────────────────────────────────────────────────────────────────────────


class MSSQLEngine(DatabaseEngine):
    """Microsoft SQL Server async engine via aioodbc + SQLAlchemy."""

    def __init__(self, config: DatabaseDefConfig):
        if not HAS_AIOODBC:
            raise RuntimeError(
                "MSSQL support requires: pip install aioodbc pyodbc. "
                "Install via: pip install axiom[mssql]"
            )

        standardized_uri = _normalize_uri(config.url)
        overflow_buffer = max(0, config.pool_max - config.pool_min)

        self.engine: Any = create_async_engine(
            standardized_uri,
            pool_size=config.pool_min,
            max_overflow=overflow_buffer,
            pool_timeout=config.connection_timeout,
            pool_recycle=config.max_lifetime,
            pool_pre_ping=False,
        )

    async def connect(self) -> None:
        """Driver pools are natively managed by the lazy-loading SQLAlchemy core."""
        pass

    async def disconnect(self) -> None:
        await self.engine.dispose()

    async def health_check(self) -> bool:
        try:
            async with self.engine.connect() as conn:
                await conn.execute(text("SELECT 1"))
            return True
        except Exception:
            return False

    async def list_tables(self) -> List[TableInfo]:
        sql = "SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = 'BASE TABLE' ORDER BY TABLE_NAME"
        async with self.engine.connect() as conn:
            result = await conn.execute(text(sql))
            return [TableInfo(name=row[0], row_count_estimate=0) for row in result]

    async def count_tables(self) -> int:
        sql = "SELECT COUNT(*) FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = 'BASE TABLE'"
        async with self.engine.connect() as conn:
            result = await conn.execute(text(sql))
            return result.scalar()

    async def describe_table(self, table: str) -> List[ColumnInfo]:
        sql = """
        SELECT
            c.COLUMN_NAME, c.DATA_TYPE, c.IS_NULLABLE,
            CASE WHEN pk.COLUMN_NAME IS NOT NULL THEN 1 ELSE 0 END AS IS_PK
        FROM INFORMATION_SCHEMA.COLUMNS c
        LEFT JOIN (
            SELECT ku.COLUMN_NAME
            FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc
            JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE ku ON tc.CONSTRAINT_NAME = ku.CONSTRAINT_NAME
            WHERE tc.CONSTRAINT_TYPE = 'PRIMARY KEY' AND ku.TABLE_NAME = :table
        ) pk ON pk.COLUMN_NAME = c.COLUMN_NAME
        WHERE c.TABLE_NAME = :table
        ORDER BY c.ORDINAL_POSITION
        """
        async with self.engine.connect() as conn:
            result = await conn.execute(text(sql), {"table": table})
            return [
                ColumnInfo(
                    name=row[0],
                    type=row[1],
                    nullable=(row[2] == "YES"),
                    primary_key=bool(row[3]),
                )
                for row in result
            ]

    async def get_foreign_keys(self, table: str) -> List[ForeignKeyInfo]:
        sql = """
        SELECT
            c.name AS column_name,
            OBJECT_NAME(fk.referenced_object_id) AS referenced_table,
            rc.name AS referenced_column
        FROM sys.foreign_keys fk
        INNER JOIN sys.foreign_key_columns fkc ON fk.object_id = fkc.constraint_object_id
        INNER JOIN sys.columns c ON fkc.parent_object_id = c.object_id AND fkc.parent_column_id = c.column_id
        INNER JOIN sys.columns rc ON fkc.referenced_object_id = rc.object_id AND fkc.referenced_column_id = rc.column_id
        WHERE OBJECT_NAME(fk.parent_object_id) = :table;
        """
        async with self.engine.connect() as conn:
            result = await conn.execute(text(sql), {"table": table})
            return [
                ForeignKeyInfo(
                    column=row[0],
                    referenced_table=row[1],
                    referenced_column=row[2],
                )
                for row in result
            ]

    async def execute(
        self,
        sql: str,
        params: Optional[Dict[str, Any]] = None,
        return_format: str = "json",
    ) -> QueryResult:
        query_params = params or {}
        statement = text(sql)

        async with self.engine.connect() as conn:
            if _is_mutation_query(sql):
                return await _execute_mutation(conn, statement, query_params)
            return await _execute_read(conn, statement, query_params, return_format)

    @property
    def dialect(self) -> str:
        return "tsql"

File: src/db/engines/mssql.rs
use async_trait::async_trait;
use std::sync::Arc;
use tiberius::{Client, Config, Query};
use tokio::net::TcpStream;
use tokio_util::compat::{TokioAsyncWriteCompatExt, Compat};
use tokio::sync::Mutex;
use serde_json::Value;

use crate::config::schema::DatabaseDefConfig;
use crate::db::engines::base::{ColumnInfo, DatabaseEngine, ForeignKeyInfo, QueryResult, TableInfo};

pub struct MssqlDatabaseEngine {
    config: DatabaseDefConfig,
    client: Option<Arc<Mutex<Client<Compat<TcpStream>>>>>,
}

impl MssqlDatabaseEngine {
    pub fn new(config: DatabaseDefConfig) -> Self {
        Self {
            config,
            client: None,
        }
    }
}

#[async_trait]
impl DatabaseEngine for MssqlDatabaseEngine {
    async fn connect(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        let config = Config::from_jdbc_string(&self.config.url)?;
        let tcp = TcpStream::connect(config.get_addr()).await?;
        tcp.set_nodelay(true)?;
        
        let client = Client::connect(config, tcp.compat_write()).await?;
        self.client = Some(Arc::new(Mutex::new(client)));
        
        Ok(())
    }

    async fn disconnect(&self) -> Result<(), Box<dyn std::error::Error>> {
        Ok(())
    }

    async fn health_check(&self) -> bool {
        if let Some(client) = &self.client {
            let mut guard = client.lock().await;
            let ok = guard.query("SELECT 1", &[]).await.is_ok(); ok
        } else {
            false
        }
    }

    async fn list_tables(
        &self,
        cursor: Option<String>,
        limit: usize,
    ) -> Result<Vec<TableInfo>, Box<dyn std::error::Error>> {
        let client_arc = self.client.as_ref().ok_or("Not connected")?;
        let mut client = client_arc.lock().await;
        
        let mut query_str = "SELECT TABLE_NAME as table_name FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = 'BASE TABLE'".to_string();
        
        if cursor.is_some() {
            query_str.push_str(" AND TABLE_NAME > @P1");
        }
        query_str.push_str(" ORDER BY TABLE_NAME ASC");
        query_str.push_str(&format!(" OFFSET 0 ROWS FETCH NEXT {} ROWS ONLY", limit));
        
        let mut query = Query::new(&query_str);
        if let Some(c) = &cursor {
            query.bind(c.clone());
        }
        
        let stream = query.query(&mut *client).await?;
        let rows = stream.into_first_result().await?;
        
        let mut tables = Vec::new();
        for row in rows {
            if let Some(name) = row.try_get::<&str, _>("table_name")? {
                tables.push(TableInfo {
                    name: name.to_string(),
                    row_count_estimate: 0,
                    columns: None,
                    foreign_keys: None,
                });
            }
        }
        Ok(tables)
    }

    async fn count_tables(&self) -> Result<i64, Box<dyn std::error::Error>> {
        let client_arc = self.client.as_ref().ok_or("Not connected")?;
        let mut client = client_arc.lock().await;
        let stream = client.query("SELECT COUNT(*) as count FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = 'BASE TABLE'", &[]).await?;
        let rows = stream.into_first_result().await?;
        
        if let Some(row) = rows.first() {
            if let Some(count) = row.try_get::<i32, _>("count")? {
                return Ok(count as i64);
            }
        }
        Ok(0)
    }

    async fn describe_table(&self, _table: &str) -> Result<Vec<ColumnInfo>, Box<dyn std::error::Error>> {
        Ok(vec![])
    }

    async fn get_foreign_keys(&self, _table: &str) -> Result<Vec<ForeignKeyInfo>, Box<dyn std::error::Error>> {
        Ok(vec![])
    }

    async fn execute(&self, sql: &str, params: &[serde_json::Value]) -> Result<QueryResult, Box<dyn std::error::Error>> {
        let client_arc = self.client.as_ref().ok_or("Not connected")?;
        let mut client = client_arc.lock().await;
        
        let is_mutation = sql.trim().to_uppercase().starts_with("INSERT")
            || sql.trim().to_uppercase().starts_with("UPDATE")
            || sql.trim().to_uppercase().starts_with("DELETE");

        // Keep allocations alive during query
        let mut string_params = Vec::new();
        let mut int_params = Vec::new();
        let mut float_params = Vec::new();
        let mut bool_params = Vec::new();
        
        for param in params {
            match param {
                Value::String(s) => string_params.push(s.clone()),
                Value::Number(n) => {
                    if let Some(i) = n.as_i64() {
                        int_params.push(i);
                    } else if let Some(f) = n.as_f64() {
                        float_params.push(f);
                    } else {
                        string_params.push(n.to_string());
                    }
                }
                Value::Bool(b) => bool_params.push(*b),
                _ => string_params.push(param.to_string()),
            }
        }

        let mut query = Query::new(sql);
        
        let mut str_idx = 0;
        let mut int_idx = 0;
        let mut float_idx = 0;
        let mut bool_idx = 0;
        
        for param in params {
            match param {
                Value::String(_) => {
                    query.bind(&string_params[str_idx]);
                    str_idx += 1;
                }
                Value::Number(n) => {
                    if n.is_i64() {
                        query.bind(int_params[int_idx]);
                        int_idx += 1;
                    } else if n.is_f64() {
                        query.bind(float_params[float_idx]);
                        float_idx += 1;
                    } else {
                        query.bind(&string_params[str_idx]);
                        str_idx += 1;
                    }
                }
                Value::Bool(_) => {
                    query.bind(bool_params[bool_idx]);
                    bool_idx += 1;
                }
                Value::Null => {
                    query.bind(None::<&str>);
                }
                _ => {
                    query.bind(&string_params[str_idx]);
                    str_idx += 1;
                }
            }
        }

        if is_mutation {
            let result = query.execute(&mut *client).await?;
            return Ok(QueryResult {
                columns: None,
                rows: None,
                // tiberius execute result gives total rows affected
                affected_rows: Some(result.total() as u64),
            });
        }

        let stream = query.query(&mut *client).await?;
        let rows = stream.into_first_result().await?;

        let mut column_names = Vec::new();
        let mut result_rows = Vec::new();

        if let Some(first) = rows.first() {
            for col in first.columns() {
                column_names.push(col.name().to_string());
            }
        }

        for row in rows {
            let mut json_obj = serde_json::Map::new();
            for col in row.columns() {
                // Simplified stringification for JSON dump, fallback to generic formatting
                let name = col.name();
                if let Ok(Some(s)) = row.try_get::<&str, _>(name) {
                    json_obj.insert(name.to_string(), Value::String(s.to_string()));
                } else if let Ok(Some(i)) = row.try_get::<i32, _>(name) {
                    json_obj.insert(name.to_string(), Value::Number(i.into()));
                } else if let Ok(Some(b)) = row.try_get::<bool, _>(name) {
                    json_obj.insert(name.to_string(), Value::Bool(b));
                } else {
                    // Very generic fallback
                    json_obj.insert(name.to_string(), Value::Null);
                }
            }
            result_rows.push(Value::Object(json_obj));
        }

        Ok(QueryResult {
            columns: Some(column_names),
            rows: Some(result_rows),
            affected_rows: None,
        })
    }

    fn dialect(&self) -> &str {
        "mssql"
    }
}

